src/duckula/avro/schema.clj

```diff
@@ -3,72 +3,88 @@
-      Keyword)
-    (org.apache.avro
-      Schema
-      Schema$Field
-      Schema$Type)))
-
+   (clojure.lang
+    Keyword)
+   (org.apache.avro
+    Schema
+    Schema$Field
+    Schema$Type)))
 
 (def ByteArray (Class/forName "[B"))
 
 (def any s/Any)
 
+(def null-val (with-meta (s/pred nil?) {:nullable? true}))
+
 (def string s/Str)
 (def any-map {Keyword any})
 
-
 (defn ->map
   [{:keys [avro-schema mangle-names?]}]
+  {:pre [avro-schema]}
   (condp = (.getType ^Schema avro-schema)
 
     ;; Primitive types
     Schema$Type/BOOLEAN s/Bool
-    Schema$Type/INT     Integer
-    Schema$Type/LONG    Long
-    Schema$Type/FLOAT   Float
-    Schema$Type/DOUBLE  Double
-    Schema$Type/BYTES   ByteArray
-    Schema$Type/STRING  s/Str
+    Schema$Type/INT s/Int
+    Schema$Type/LONG s/Int
+    Schema$Type/FLOAT Float
+    Schema$Type/DOUBLE Double
+    Schema$Type/BYTES ByteArray
+    Schema$Type/STRING s/Str
 
     ;; Complex Types
     Schema$Type/RECORD
     (apply hash-map (mapcat (fn [key val]
-                              (let [k (if mangle-names?
+                              (let [{:keys [nullable?]} (meta val)
+                                    k (if mangle-names?
                                         (-> key (str/replace \_ \-) keyword)
                                         (keyword key))]
-                                [k val]))
+                                (if nullable?
+                                  [(s/optional-key k) val]
+                                  [k val])))
                             (map #(.name ^Schema$Field %) (.getFields ^Schema avro-schema))
                             (map #(->map {:avro-schema (.schema ^Schema$Field %)
                                           :mangle-names? mangle-names?})
                                  (.getFields ^Schema avro-schema))))
 
+    ;; enums, but accept both :keywords and strings - like Avro layer
     Schema$Type/ENUM
-    (apply s/enum (map (fn [enum]
-                         (if mangle-names?
-                           (str/replace enum \_ \-)
-                           enum))
-                       (.getEnumSymbols ^Schema avro-schema)))
+    (apply s/enum (mapcat (fn [enum]
+                            (let [enum-val-str (if mangle-names?
+                                                 (str/replace enum \_ \-)
+                                                 enum)]
+                              [enum-val-str (keyword enum-val-str)]))
+                          (.getEnumSymbols ^Schema avro-schema)))
 
+    ;; arrays
     Schema$Type/ARRAY
     [(->map {:avro-schema (.getElementType ^Schema avro-schema) :mangle-names? mangle-names?})]
 
+    ;; maps, but accept both strings and keywords as keys
     Schema$Type/MAP
-    {s/Str (->map {:avro-schema  (.getValueType ^Schema avro-schema) :mangle-names? mangle-names?})}
+    {(s/cond-pre s/Str Keyword) (->map {:avro-schema (.getValueType ^Schema avro-schema) :mangle-names? mangle-names?})}
```

Comment:
This means all maps will accept keywords or strings as keys - is that what we want?

Reply:
It's confusing indeed, probably because of how Abracad is designed, see here: 

https://github.com/nomnom-insights/abracad#basic-deserialization specifically:  

> The Avro specification allows field names to contain the _ character but disallows the - character. Clojure keywords frequently contain - but rarely contain _. Abracad attempts to work around this difference by mapping _ in Avro field names to - in Clojure symbols and vice-versa. The current implementation of this conversion does not handle keywords containing _ instead, which is probably a bug. This mangling may be disabled by binding abracad.avro.util/*mangle-names* to false.

In retrospect, maybe we should remove name mangling entirely in our Abracad fork (along with other features) , on the Avro layer, but have a middleware for Duckula to convert between kebab and snake case on the request path, and snake case to kebab on the response path.

I tried to give it a better name when using in the Duckula config: https://github.com/nomnom-insights/nomnom.duckula/blob/34107b2927f3a5a3db45b0947cc9bd8f66418d60/src/duckula/handler.clj#L11-L15 

but of course our internal services predate that. And the config option name is not 100% clear either - maybe it should be `kebab-case-keywords-and-keys?` or something like that.